host_core/lib/host_core/control_interface/server.ex

```diff
@@ -229,45 +290,59 @@ defmodule HostCore.ControlInterface.Server do
   # Auction Actor
   # input: #{"actor_ref" => "...", "constraints" => %{}}
   defp handle_request({"auction", "actor"}, body, _reply_to) do
-    auction_request = Jason.decode!(body)
-    host_labels = HostCore.Host.host_labels()
-    required_labels = auction_request["constraints"]
-
-    if Map.equal?(host_labels, Map.merge(host_labels, required_labels)) do
-      ack = %{
-        actor_ref: auction_request["actor_ref"],
-        constraints: auction_request["constraints"],
-        host_id: HostCore.Host.host_key()
-      }
-
-      {:reply, Jason.encode!(ack)}
+    with {:ok, auction_request} <- Jason.decode(body),
+         true <-
+           ["constraints", "actor_ref"]
+           |> Enum.all?(&Map.has_key?(auction_request, &1)) do
+      host_labels = HostCore.Host.host_labels()
+      required_labels = auction_request["constraints"]
+
+      if Map.equal?(host_labels, Map.merge(host_labels, required_labels)) do
+        ack = %{
+          actor_ref: auction_request["actor_ref"],
+          constraints: auction_request["constraints"],
+          host_id: HostCore.Host.host_key()
+        }
+
+        {:reply, Jason.encode!(ack)}
+      else
+        # We don't respond to an auction request if this host cannot satisfy the constraints
+        :ok
+      end
     else
-      # We don't respond to an auction request if this host cannot satisfy the constraints
-      :ok
+      _ ->
+        {:reply, failure_ack("Invalid JSON request for actor auction")}
     end
   end
 
   # Auction Provider
   # input: #{"actor_ref" => "...", "constraints" => %{}}
   defp handle_request({"auction", "provider"}, body, _reply_to) do
-    auction_request = Jason.decode!(body)
-    host_labels = HostCore.Host.host_labels()
-    required_labels = auction_request["constraints"]
-
-    # TODO - don't answer this request if we're already running a provider
-    # that matches this link_name and ref.
-    if Map.equal?(host_labels, Map.merge(host_labels, required_labels)) do
-      ack = %{
-        provider_ref: auction_request["provider_ref"],
-        link_name: auction_request["link_name"],
-        constraints: auction_request["constraints"],
-        host_id: HostCore.Host.host_key()
-      }
-
-      {:reply, Jason.encode!(ack)}
+    with {:ok, auction_request} <- Jason.decode(body),
+         true <-
+           ["constraints", "provider_ref"]
```

Comment:
I went back and forth on if `link_name` should be here as well, initially I thought it should but if my assumptions are correct it shouldn't be needed:
1. `auction_request["link_name"]` without one present resolves to `nil`
1. `link_name` is optional on the control interface, so being specified as `link_name: nil` here simply ends up being the equivalent of `None` in Rust

Reply:
I actually had intended it to be there but I missed it. Instead I think I'll make it default to `default` rather than nil.